crates/services/sync/src/import.rs

```diff
@@ -270,50 +329,61 @@ fn get_block_stream<
     P: PeerToPeerPort + Send + Sync + 'static,
     C: ConsensusPort + Send + Sync + 'static,
 >(
+    peer_id: PeerId,
     range: RangeInclusive<u32>,
     params: &Config,
     p2p: Arc<P>,
     consensus: Arc<C>,
-) -> impl Stream<Item = impl Future<Output = anyhow::Result<Option<(PeerId, SealedBlock)>>>>
-{
-    get_header_stream(range, params, p2p.clone()).map({
-        let p2p = p2p.clone();
-        let consensus_port = consensus.clone();
-        move |batch| {
-            {
-                let p2p = p2p.clone();
-                let consensus_port = consensus_port.clone();
-                get_sealed_blocks(batch, p2p.clone(), consensus_port.clone())
-            }
-            .instrument(tracing::debug_span!("consensus_and_transactions"))
-            .in_current_span()
-        }
-    })
+) -> impl Stream<Item = impl Future<Output = (Vec<SealedBlock>, Option<ImportError>)>> {
+    let Config {
+        header_batch_size, ..
+    } = params;
+    let header_stream = get_header_stream(peer_id.clone(), range, params, p2p.clone());
+    let generator =
+        futures::stream::repeat((peer_id.clone(), p2p.clone(), consensus.clone()));
+    let iter = header_stream.zip(generator.clone());
+    let checked_headers = iter
+        .then(|(header, (peer_id, p2p, consensus))| async move {
+            let header = header?;
+            check_sealed_header(&header, peer_id, p2p.as_ref(), consensus.as_ref())
+                .await?;
+            await_da_height(&header, consensus.as_ref()).await?;
+            Ok(header)
+        })
+        .into_scan_err()
+        .scan_err();
+    checked_headers
+        .chunks(*header_batch_size as usize)
+        .zip(generator)
+        .map(|(headers, (peer_id, p2p, ..))| get_blocks(p2p, peer_id, headers))
```

Comment:
Yep, that makes sense. I will try that.

Reply:
I've updated the PR to use the concept of a `Batch`. A `Batch` is simply a collection of `T` holding all the successful items returned by a batch request. A `Batch` is in an error state if the number of `T` it holds is less than the amount we requested. 
Now, we operate on `Batches` rather than flattening any `Vec`.